csrc/id_model/id_graphs.cpp

```diff
@@ -0,0 +1,1768 @@
+
+    VectorOfUniqueEntries<IterDomain*> resolved_ids =
+        exact_resolution_group->intersect(*loop_group);
+    auto promoted_iel_groups =
+        intersection_exact_loop_graph.toGroups(resolved_ids);
+
+    if (promoted_iel_groups.size() == 0) {
+      continue;
+    }
+
+    if (promoted_iel_groups.size() > 1) {
+      std::stringstream err_msg;
+
+      err_msg
+          << "Invalid multiple broadcast resolution within shared loops detected, group:\n  "
+          << iel_group->toString() << "\nIs being broadcasted to:";
+
+      for (auto entry : promoted_iel_groups) {
+        err_msg << "\n  " << entry->toString();
+      }
+      TORCH_INTERNAL_ASSERT(false, err_msg.str());
+    }
+
+    iel_promotion_map[iel_group] = promoted_iel_groups.front()->front();
+  }
+
+  for (auto iel_group :
+       intersection_exact_loop_graph.disjointIdSets().disjointSets()) {
+    auto entry_it = iel_promotion_map.find(iel_group);
+    if (entry_it == iel_promotion_map.end()) {
+      continue;
+    }
+  }
+
+  IdGraphStmtSort iel_stmt_sort(intersection_exact_loop_graph);
+
+  for (auto iel_expr : iel_stmt_sort.exprs()) {
+    auto input_groups = intersection_exact_loop_graph.inputGroups(iel_expr);
+
+    // Check if any inputs need promotion indicating this expr group needs to
+    // be replayed with promoted inputs
+    std::vector<IterDomain*> promoted_inputs;
+    bool an_input_was_promoted = false;
+
+    for (auto inp : input_groups) {
+      auto inp_promo_it = iel_promotion_map.find(inp);
+      if (inp_promo_it == iel_promotion_map.end()) {
+        promoted_inputs.push_back(inp->front());
+      } else {
+        promoted_inputs.push_back(inp_promo_it->second);
+        an_input_was_promoted = true;
+      }
+    }
+
+    if (!an_input_was_promoted) {
+      // No inputs need promotion so just continue
+      continue;
+    }
+
+    Expr* replay = nullptr;
+
+    IdGroups promoted_input_groups;
+    for (auto inp_id : promoted_inputs) {
+      if (intersection_exact_loop_graph.hasGroup(inp_id)) {
+        promoted_input_groups.pushBack(
+            intersection_exact_loop_graph.toGroup(inp_id));
+      }
+    }
+
+    // Before replaying, check if there's already an expression like this, if so
+    // use that for promotion. We would need the iel entries for non-promoted
+    // inputs to match exactly to reuse the expression.
+    //
+    // Unfortunately this doesn't actually seem to save any replays because
+    // we're not adding the replayed expression to the iel graph since we're
+    // traversing the iel graph.
+    //
+    // TODO: Can we reduce the number of new expressions generated here?
+    ExprGroups non_promoted_input_uses;
+    for (auto iel_group : promoted_input_groups.intersect(input_groups)) {
+      non_promoted_input_uses.pushBack(
+          intersection_exact_loop_graph.getUniqueUses(iel_group));
+    }
+
+    for (auto iel_use_group : non_promoted_input_uses) {
```

Comment:
For single-input exprs such as `split`, there can be no non-promoted input, so it's always replayed. Is it intended? I wonder why the use expr of the promote target ID can be reused.

For example, when we have a merge and then a split, and the one of the merge input is promoted, the merge expr with the promotion target ID is going to be reused instead of creating a new expr. Then, for the split, it's always replayed. Why can't we reuse the split too?

Reply:
```// TODO: Can we reduce the number of new expressions generated here?```
I’m pretty confident we could reuse more expressions, but I think I simply started prioritizing correctness rather than optimization.